sdk/ml/azure-ai-ml/azure/ai/ml/entities/_component/_additional_includes.py

```diff
@@ -0,0 +1,402 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import os
+import shutil
+import tempfile
+import zipfile
+from concurrent.futures import ThreadPoolExecutor
+from multiprocessing import cpu_count
+from pathlib import Path
+from typing import Union, List
+import yaml
+
+from azure.ai.ml.entities._validation import MutableValidationResult, _ValidationResultBuilder
+from azure.ai.ml.constants._common import AzureDevopsArtifactsType
+
+from .code import ComponentIgnoreFile
+from ._artifact_cache import ArtifactCache
+from ..._utils.utils import is_concurrent_component_registration_enabled, is_private_preview_enabled
+from ..._utils._asset_utils import traverse_directory, IgnoreFile
+from ...entities._util import _general_copy
+
+
+PLACEHOLDER_FILE_NAME = "_placeholder_spec.yaml"
+ADDITIONAL_INCLUDES_KEY = "additional_includes"
+ADDITIONAL_INCLUDES_SUFFIX = ".additional_includes"
+
+
+class AdditionalIncludes:
+    def __init__(self, code_path: Union[None, str], yaml_path: str, additional_includes: List[Union[str, dict]] = None):
+        self._yaml_path = yaml_path
+        self._code_path = code_path
+        self._additional_includes = additional_includes if additional_includes else []
+
+        self._tmp_code_path = None
+        self._includes = None
+        # artifact validation is done on loading now, so need a private variable to store the result
+        self._artifact_validate_result = None
+
+    @property
+    def additional_includes(self):
+        return self._additional_includes
+
+    @property
+    def code_path(self) -> Union[None, Path]:
+        if self._code_path is not None:
+            return (self.yaml_path.parent / self._code_path).resolve()
+        return self._code_path
+
+    @property
+    def code(self):
+        return self._tmp_code_path if self._tmp_code_path else self.code_path
+
+    @property
+    def yaml_path(self) -> Path:
+        if self._yaml_path is None:
+            # if yaml path is not specified, take working directory as root folder and use a not created temp file name
+            return Path.cwd() / PLACEHOLDER_FILE_NAME
+        return Path(self._yaml_path)
+
+    @property
+    def additional_includes_file_path(self) -> Path:
```

Comment:
as offline synced, maybe use `base_path` in this class instead.

Reply:
updated